### DISCUSSION TITLE
Efficiently getting a nodes set of ancestors

### DISCUSSION THREAD

**Discussion by lewis6991** [tree-sitter/tree-sitter#3716] [Ideas]
Due to the way treesitter internal structure is designed, traversing up a tree is very expensive, as the internal structures have no information about what other nodes they are children of. Therefore in order to determine the parent of a node, you need to search for such node from the root of the tree, and from there you can figure out a nodes line of ancestors.

Here are a few methods I tried to do this efficiently.

Note: this was done for Neovim, so the implementation are designed for the Lua bindings.

### `ts_node_parent()` in a loop

Very inefficient, as each loop iteration does a fresh search from the root.

Complexity: $O(n^2)$

<details>
<summary>Details</summary>

```lua
--- @param n TSNode
--- @return TSNode[]
local function get_ancestors(n)
  local ret = {}
  while n do
    ret[#ret + 1] = n
    n = n:parent()
  end
  return ret
end
```

</details>

### `ts_node_child_containing_descendant()` in a loop

Better, but still inefficient.

Complexity: $O(n^2)$

<details>
<summary>Details</summary>

```lua
--- @param n TSNode
--- @return TSNode[]
local function get_ancestors(n)
  local ret = {}
  local p = dec:tree():root()
  while p do
    table.insert(ret, 1, p)
    p = p:child_containing_descendant(n)
  end
  return ret
end
```

</details>

### Custom function using the tree cursor public API

Only one search is done so _should_ be efficient.

Complexity: $O(n)$

<details>
<summary>Details</summary>

Lua binding
```c
static bool navigate_to_node(TSTreeCursor *c, TSNode n)
{
  uint32_t start_byte = ts_node_start_byte(n);

  while (true) {
    TSNode cur = ts_tree_cursor_current_node(c);

    if (ts_node_eq(cur, n)) {
      return true;
    }

    if (ts_tree_cursor_goto_first_child_for_byte(c, start_byte) >= 0
       || ts_tree_cursor_goto_next_sibling(c)) {
      continue;
    }

    break;
  }

  return false;
}

static int node_ancestors(lua_State *L)
{
  TSNode node = node_check(L, 1);
  lua_newtable(L);

  TSNode root = ts_tree_root_node(node.tree);
  TSTreeCursor c = ts_tree_cursor_new(root);

  int curr_index = 0;
  if (navigate_to_node(&c, node)) {
    // Found the descendant, traverse back up and collect all the parents
    while (ts_tree_cursor_goto_parent(&c)) {
      push_node(L, ts_tree_cursor_current_node(&c), 1);
      lua_rawseti(L, -2, ++curr_index);
    }
  }

  ts_tree_cursor_delete(&c);

  return 1;
}

```

</details>

### Custom treesitter function using internal API's (`ts_node_ancestors`).

Only one search, uses internal API's, should be quick.

Complexity: $O(n)$

<details>
<summary>Details</summary>

added to `lib/src/node.c`
```c
TSNode* ts_node_ancestors(TSNode self, uint32_t* length) {
  TSNode node = ts_tree_root_node(self.tree);
  Array(TSNode) array = array_new();
  array_push(&array, node);

  uint32_t end_byte = ts_node_end_byte(self);
  if (node.id == self.id) return NULL;

  bool did_descend = true;
  while (did_descend) {
    did_descend = false;

    TSNode child;
    NodeChildIterator iterator = ts_node_iterate_children(&node);
    while (ts_node_child_iterator_next(&iterator, &child)) {
      if (
        ts_node_start_byte(child) > ts_node_start_byte(self) ||
        child.id == self.id
      ) break;
      if (iterator.position.bytes >= end_byte && ts_node_child_count(child) > 0) {
        node = child;
        if (ts_node__is_relevant(child, true)) {
          array_push(&array, node);
        }
        did_descend = true;
        break;
      }
    }
  }

  *length = array.size;
  return array.contents;
}
```

Lua binding
```c
static int node_ancestors(lua_State *L)
{
  TSNode node = node_check(L, 1);
  lua_newtable(L);

  uint32_t length;
  TSNode *nodes = ts_node_ancestors(node, &length);

  for (uint32_t i = 0; i < length; i++) {
    push_node(L, nodes[i], 1);
    lua_rawseti(L, -2, (int)(i + 1));
  }

  xfree(nodes);

  return 1;
}
```

</details>

## Results
```
ts_node_parent.                      0.011738
ts_child_node_containing_descendant  0.00788
tree_cursor_public                   - (similar to ts_child_node_containing_descendant)
custom_private                       0.00304
```

Overall `ts_child_node_containing_descendant()` isn't too much slower than the custom function that does a single search, however the test was done are a relatively normal file. For very large files, I suspect the custom function will scale much better.

The tree cursor appears to be very slow for this application. Not sure why.

EDIT: modified the tree cursor implementation to also use `ts_tree_cursor_goto_first_child_for_byte()`, now it appears to be faster than `ts_child_node_containing_descendant`

**Comment by lewis6991** [comment]
I also attempted to draft a PR that added a reference into the nodes subtree (#3714), however that isn't possible since subtrees can be re-used in incremental parsing.

**Comment by vanaigr** [comment]
> ### `ts_node_child_containing_descendant()` in a loop
> Better, but still inefficient.
> 
> Complexity: O(_n<sup>2</sup>_)

Is looping with `ts_node_child_containing_descendant()` quadratic on the depth of the tree? I know using it in a loop currently (e445532a1fea3b1dda93cee61c534f5b9acc9c16) becomes quadratic when searching for a zero-width node, but otherwise, it should be linear. At least, that was the intention.


Also, picking the first child

> ```c
> static bool navigate_to_node(TSTreeCursor *c, TSNode n)
> {
>   uint32_t start_byte = ts_node_start_byte(n);
> 
>   while (true) {
>     TSNode cur = ts_tree_cursor_current_node(c);
> 
>     if (ts_node_eq(cur, n)) {
>       return true;
>     }
> 
>     if (ts_tree_cursor_goto_first_child_for_byte(c, start_byte) >= 0
>        || ts_tree_cursor_goto_next_sibling(c)) {
>       continue;
>     }
> 
>     break;
>   }
> 
>   return false;
> }
> ```

 might not work for zero-width nodes, e.g. for siblings [0,10) [10,20), a zero-width child node at [10,10) can be in either of the parents (hence the O(_n<sup>2</sup>_) fallback in `ts_node_child_containing_descendant()`).

Ideally, the tree cursor would have a method for going to a given node in O(_n_) time even for a zero-width node (since it is possible to avoid  O(_n<sup>2</sup>_) if you keep an array of parents).
